src/RazorSdk/SourceGenerators/StaticCompilationTagHelperFeature.cs

```diff
@@ -26,6 +26,7 @@ public IReadOnlyList<TagHelperDescriptor> GetDescriptors()
 
             var context = TagHelperDescriptorProviderContext.Create(results);
             context.SetCompilation(Compilation);
+            context.Items.SetTagHelperDiscoveryFilter(DiscoveryFilter);
```

Comment:
This is the important bit. In the absence of this, the providers were repeatedly scanning everything and caching duplicate values.